### PR TITLE
Add a function to calculate Cumulants

### PR DESCRIPTION
**Pull request by JaydevSR**
I was working with cumulants so though it would be a nice addition to this library. I have added a function to calculate the cumulants of a given vector using a recursive definition (see: https://doi.org/10.2307/2684642). I have also added some tests.

### CODE REVIEW

File: src/moments.jl [hunk 1]
```diff
@@ -363,3 +363,44 @@ end
...
 kurtosis(v::RealArray) = kurtosis(v, mean(v))
 kurtosis(v::RealArray, wv::AbstractWeights) = kurtosis(v, wv, mean(v, wv))
+
+"""
+    cumulant(v, k, [wv::AbstractWeights], m=mean(v))
+
+Return the `k`th order cumulant of a real-valued array `v`, optionally
+specifying a weighting vector `wv` and a center `m`. This quantity is calculated using a recursive definition on lower order cumulants and central moments.
+"""
+function cumulant(v::RealArray, k::Int, m::Real)
```
Comment (nalimilan) [inline src/moments.jl:373]:
```suggestion
function cumulant(v::RealArray, k::Int, m::Real=mean(v))
```

File: src/moments.jl [hunk 2]
```diff
@@ -363,3 +363,44 @@ end
...
+        @inbounds kn = cmoms[i]
+        for j=2:i-2
+            @inbounds kn -= binomial(i-1, j)*cmoms[j]*cumls[i-j]
+        end
+        @inbounds cumls[i] = kn
+    end
+    return cumls[k]
+end
+
+function cumulant(v::RealArray, k::Int, wv::AbstractWeights, m::Real)
```
Comment (nalimilan) [inline src/moments.jl:389]:
```suggestion
function cumulant(v::RealArray, k::Int, wv::AbstractWeights, m::Real=mean(v, wv))
```

File: src/moments.jl [hunk 3]
```diff
@@ -363,3 +363,44 @@ end
...
+        for j=2:i-2
+            @inbounds kn -= binomial(i-1, j)*cmoms[j]*cumls[i-j]
+        end
+        @inbounds cumls[i] = kn
+    end
+    return cumls[k]
+end
+
+cumulant(v::RealArray, k::Int) = cumulant(v, k, mean(v))
+cumulant(v::RealArray, k::Int, wv::AbstractWeights) = cumulant(v, k, wv::RealArray, mean(v, wv))
```
Comment (nalimilan) [inline src/moments.jl:406]:
```suggestion
```

File: src/moments.jl [hunk 4]
```diff
@@ -363,3 +363,44 @@ end
...
+    cmoms[1] = 0
+    cumls[1] = m
+    for i=2:k
+        @inbounds cmoms[i] =  moment(v, i, m)
+        @inbounds kn = cmoms[i]
+        for j=2:i-2
+            @inbounds kn -= binomial(i-1, j)*cmoms[j]*cumls[i-j]
+        end
+        @inbounds cumls[i] = kn
+    end
```
Comment (nalimilan) [inline src/moments.jl:385]: `@inbounds` is probably not worth it here given that most of the work will happen in the call to `moment`:
```suggestion
    for i = 2:k
        kn = moment(v, i, m)
        for j = 2:i-2
             kn -= binomial(i-1, j)*cmoms[j]*cumls[i-j]
        end
        cumls[i] = kn
    end
```

File: src/moments.jl [hunk 5]
```diff
@@ -363,3 +363,44 @@ end
...
+
+"""
+    cumulant(v, k, [wv::AbstractWeights], m=mean(v))
+
+Return the `k`th order cumulant of a real-valued array `v`, optionally
+specifying a weighting vector `wv` and a center `m`. This quantity is calculated using a recursive definition on lower order cumulants and central moments.
+"""
+function cumulant(v::RealArray, k::Int, m::Real)
+    cmoms = zeros(Float64, k)
+    cumls = zeros(Float64, k)
```
Comment (nalimilan) [inline src/moments.jl:375]: Shouldn't hardcode `Float64` here:
```suggestion
    cmoms = zeros(typeof(m), k)
    cumls = zeros(typeof(m), k)
```
Could you also add a test with a non-`Float64` input type (e.g. `Int`)?

File: src/moments.jl [hunk 6]
```diff
@@ -363,3 +363,44 @@ end
 
 kurtosis(v::RealArray) = kurtosis(v, mean(v))
 kurtosis(v::RealArray, wv::AbstractWeights) = kurtosis(v, wv, mean(v, wv))
+
+"""
+    cumulant(v, k, [wv::AbstractWeights], m=mean(v))
+
+Return the `k`th order cumulant of a real-valued array `v`, optionally
+specifying a weighting vector `wv` and a center `m`. This quantity is calculated using a recursive definition on lower order cumulants and central moments.
```
Comment (nalimilan) [inline src/moments.jl:371]:
```suggestion
specifying a weighting vector `wv` and a pre-computed mean `m`.

This quantity is calculated using a recursive definition on lower-order
cumulants and central moments.
```

File: src/moments.jl [hunk 7]
```diff
@@ -363,3 +363,44 @@ end
...
+    cumls[1] = m
+    for i=2:k
+        @inbounds cmoms[i] =  moment(v, i, wv, m)
+        @inbounds kn = cmoms[i]
+        for j=2:i-2
+            @inbounds kn -= binomial(i-1, j)*cmoms[j]*cumls[i-j]
+        end
+        @inbounds cumls[i] = kn
+    end
+    return cumls[k]
```
Comment (nalimilan) [inline src/moments.jl:402]: This code is identical to the function above, so you could avoid the duplication by passing `UnitWeights` as `wv` for the unweighted case.

File: src/StatsBase.jl [hunk 8]
```diff
@@ -69,6 +69,7 @@ export
     mean_and_var,   # (mean, var)
     mean_and_std,   # (mean, std)
     mean_and_cov,   # (mean, cov)
+    cumulant,       # cumulant of given order
```
Comment (nalimilan) [inline src/StatsBase.jl:72]:
```suggestion
    cumulant,       # cumulant of given order
    mean_and_var,   # (mean, var)
    mean_and_std,   # (mean, std)
    mean_and_cov,   # (mean, cov)
```

File: docs/src/scalarstats.md [hunk 9]
```diff
@@ -35,6 +35,11 @@ kurtosis
 moment
 ```
 
+## Cumulants
+```@docs
+cumulant
+```
```
Comment (nalimilan) [inline docs/src/scalarstats.md:41]: Maybe add this to the previous section, renaming it to "Moments and cumulants"?

File: src/moments.jl [hunk 10]
```diff
@@ -363,3 +363,44 @@ end
 
 kurtosis(v::RealArray) = kurtosis(v, mean(v))
 kurtosis(v::RealArray, wv::AbstractWeights) = kurtosis(v, wv, mean(v, wv))
+
+"""
+    cumulant(v, k, [wv::AbstractWeights], m=mean(v))
```
Comment (nalimilan) [inline src/moments.jl:368]: Should we allow passing `corrected` here too? This should probably be mentioned anyway.

**Review by nalimilan — COMMENTED** [review]
Thanks, that sounds like a useful addition.

File: src/moments.jl (hunk 1 above)
Comment (devmotion) [reply src/moments.jl:373]: Maybe `m` could even be a keyword argument (probably with a more descriptive name such as `mean`)?

Comment (nalimilan) [reply src/moments.jl:373]: Yeah that would be better, but `moment` currently has a positional argument (remnants from pre-Julia 1.0 times), so we should change both at the same time.

**Comment by bkamins** [comment]
I think it is a useful addition (I have not checked the logic of code as I assume @nalimilan did 😄)

**Comment by nalimilan** [comment]
> One more thing was as all the lower order cumulants are already calculated in any computation of say `k` order cumulant, the function can also return all the cumulants up to order `k` without any extra computation. But I am not sure how to provide this functionality that fits with this library, as `moments` has no analogous method to this.

We could allow passing a range of orders, in which case a vector of cumulants would be returned.

File: src/moments.jl (hunk 10 above)
Comment (JaydevSR) [reply src/moments.jl:368]: Can you please elaborate, as moment function does not take this argument so how will it be used?

Comment (nalimilan) [reply src/moments.jl:368]: You can check whether `wv isa UnitWeights` and if so just call `moment` without weights.

File: src/moments.jl [hunk 11]
```diff
@@ -363,3 +363,40 @@ end
...
+
+Return the `k`th order cumulant of a real-valued array `v`, optionally
+specifying a weighting vector `wv` and a pre-computed mean `m`. If `k` is a `UnitRange`, then return all the cumulants of orders in this range as a vector.
```
Comment (nalimilan) [inline src/moments.jl:371]: AFAICT any range of integers works, no reason to require a `UnitRange`.
```suggestion
specifying a weighting vector `wv` and a pre-computed mean `m`.

If `k` is a range of `Integer`s, then return all the cumulants of orders in this range as a vector.
```

File: src/moments.jl [hunk 12]
```diff
@@ -363,3 +363,40 @@ end
...
+
+This quantity is calculated using a recursive definition on lower-order cumulants and central moments.
+"""
+function cumulant(v::RealArray, krange::UnitRange{Int}, wv::AbstractWeights, m::Real=mean(v, wv))
```
Comment (nalimilan) [inline src/moments.jl:375]:
```suggestion
function cumulant(v::RealArray, krange::AbstractRange{<:Integer}, wv::AbstractWeights,
                  m::Real=mean(v, wv))
```

File: src/moments.jl [hunk 13]
```diff
@@ -363,3 +363,40 @@ end
...
+    if minimum(krange) <= 0
+        throw(ArgumentError("The cumulant order must be positive."))
```
Comment (nalimilan) [inline src/moments.jl:377]:
```suggestion
        throw(ArgumentError("Cumulant orders must be strictly positive."))
```

File: src/moments.jl [hunk 14]
```diff
@@ -363,3 +363,40 @@ end
...
+    cmoms = zeros(typeof(m), k)
+    cumls = zeros(typeof(m), k)
+    cmoms[1] = 0
+    cumls[1] = m
+    for i = 2:k
+        if wv isa UnitWeights
+            kn =  moment(v, i, m)
+        else
+            kn =  moment(v, i, wv, m)
+        end
```
Comment (nalimilan) [inline src/moments.jl:389]:
```suggestion
        kn = wv isa UnitWeights ? moment(v, i, m) : moment(v, i, wv, m)
```

File: src/moments.jl [hunk 15]
```diff
@@ -363,3 +363,40 @@ end
...
+        cmoms[i] = kn
+        for j = 2:i-2
+            kn -= binomial(i-1, j)*cmoms[j]*cumls[i-j]
+        end
+        cumls[i] = kn
+    end
+    return cumls[krange]
+end
+
+cumulant(v::RealArray, krange::UnitRange{Int}, m::Real=mean(v)) = cumulant(v, krange, uweights(typeof(m), length(v)), m)
```
Comment (nalimilan) [inline src/moments.jl:399]:
```suggestion
cumulant(v::RealArray, krange::AbstractRange{<:Integer}, m::Real=mean(v)) =
    cumulant(v, krange, uweights(typeof(m), length(v)), m)
```

File: src/moments.jl [hunk 16]
```diff
@@ -363,3 +363,40 @@ end
...
+
+cumulant(v::RealArray, k::Int, wv::AbstractWeights, m::Real=mean(v, wv)) = first(cumulant(v, k:k, wv, m))
+cumulant(v::RealArray, k::Int, m::Real=mean(v)) = cumulant(v, k, uweights(typeof(m), length(v)), m)
```
Comment (nalimilan) [inline src/moments.jl:402]: I think you can simply drop these definitions and instead use a wider signature above: `krange::Union{Integer, AbstractRange{<:Integer}}`. This should work as numbers behave as single-element collections.

File: src/moments.jl (hunk 10 above)
Comment (nalimilan) [reply src/moments.jl:368]: Sorry, I didn't address the question. I thought `moment` took `corrected` like `var`, but that's not the case so everything is fine.

File: test/moments.jl [hunk 17]
```diff
@@ -290,4 +290,42 @@ end
     @test moment(x, 5, w) ≈ sum((x2 .- 4).^5) / 5
 end
 
+@testset "Cumulants with $f" for f in weight_funcs
+    x = collect(2.0:8.0)
+    @test cumulant(x, 2) ≈ moment(x, 2)
+    @test cumulant(x, 3) ≈ moment(x, 3)
+    @test cumulant(x, 4) ≈ moment(x, 4) - 3*moment(x, 2)^2
+    @test cumulant(x, 5) ≈ moment(x, 5) - 10*moment(x, 3)*moment(x, 2)
+    @test cumulant(x, 6) ≈ moment(x, 6) - 15*moment(x, 4)*moment(x, 2) - 10*moment(x, 3)^2 + 30*moment(x, 2)^3
```
Comment (nalimilan) [inline test/moments.jl:299]:
```suggestion
    @test cumulant(x, 6) ≈
        moment(x, 6) - 15*moment(x, 4)*moment(x, 2) - 10*moment(x, 3)^2 + 30*moment(x, 2)^3
```

File: test/moments.jl [hunk 18]
```diff
@@ -290,4 +290,42 @@ end
...
+
+    @test cumulant(x, 1:6) == [cumulant(x, i) for i in 1:6]
+
+    @test cumulant(x, 2, 4.0) ≈ moment(x, 2, 4.0)
+    @test cumulant(x, 3, 4.0) ≈ moment(x, 3, 4.0)
+    @test cumulant(x, 4, 4.0) ≈ moment(x, 4, 4.0) - 3*moment(x, 2, 4.0)^2
+    @test cumulant(x, 5, 4.0) ≈ moment(x, 5, 4.0) - 10*moment(x, 3, 4.0)*moment(x,2, 4.0)
+    @test cumulant(x, 6, 4.0) ≈ moment(x, 6, 4.0) - 15*moment(x, 4, 4.0)*moment(x,2, 4.0) - 10*moment(x, 3, 4.0)^2 + 30*moment(x, 2, 4.0)^3
```
Comment (nalimilan) [inline test/moments.jl:307]:
```suggestion
    @test cumulant(x, 6, 4.0) ≈
        moment(x, 6, 4.0) - 15*moment(x, 4, 4.0)*moment(x,2, 4.0) -
        10*moment(x, 3, 4.0)^2 + 30*moment(x, 2, 4.0)^3
```

File: test/moments.jl [hunk 19]
```diff
@@ -290,4 +290,42 @@ end
...
+
+    @test cumulant(x, 1:6, 4.0) == [cumulant(x, i, 4.0) for i in 1:6]
+
+    w1 = f([1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0])
+    x2 = collect(2.0:6.0)
+    @test cumulant(x, 2, w) ≈ moment(x2, 2)
+    @test cumulant(x, 3, w) ≈ moment(x2, 3)
+    @test cumulant(x, 4, w) ≈ moment(x2, 4) - 3*moment(x2, 2)^2
+    @test cumulant(x, 5, w) ≈ moment(x2, 5) - 10*moment(x2,3)*moment(x2,2)
+    @test cumulant(x, 6, w) ≈ moment(x2, 6) - 15*moment(x2,4)*moment(x2,2) + 10*moment(x2,3)^2 + 30*moment(x2,2)^3
```
Comment (nalimilan) [inline test/moments.jl:317]:
```suggestion
    @test cumulant(x, 6, w) ≈
        moment(x2, 6) - 15*moment(x2,4)*moment(x2,2) + 10*moment(x2,3)^2 + 30*moment(x2,2)^3
```

File: test/moments.jl [hunk 20]
```diff
@@ -290,4 +290,42 @@ end
...
+
+    @test cumulant(x, 1:6, w) == [cumulant(x2, i) for i in 1:6]
+
+    x3 = collect(2:8)
+    @test cumulant(x3, 6) ≈ moment(x3, 6) - 15*moment(x3, 4)*moment(x3, 2) - 10*moment(x3, 3)^2 + 30*moment(x3, 2)^3
```
Comment (nalimilan) [inline test/moments.jl:322]:
```suggestion
    @test cumulant(x3, 6) ≈
        moment(x3, 6) - 15*moment(x3, 4)*moment(x3, 2) - 10*moment(x3, 3)^2 +
        30*moment(x3, 2)^3
```

File: test/moments.jl [hunk 21]
```diff
@@ -290,4 +290,42 @@ end
...
+
+    w2 = f([1, 1, 1, 1, 1, 0, 0])
+    x4 = collect(2:6)
+    @test cumulant(x3, 6, w) ≈ moment(x4, 6) - 15*moment(x4, 4)*moment(x4, 2) + 10*moment(x4, 3)^2 + 30*moment(x4, 2)^3
```
Comment (nalimilan) [inline test/moments.jl:326]:
```suggestion
    @test cumulant(x3, 6, w) ≈
        moment(x4, 6) - 15*moment(x4, 4)*moment(x4, 2) +
        10*moment(x4, 3)^2 + 30*moment(x4, 2)^3
```

**Review by nalimilan — COMMENTED** [review]
Thanks, looks good!

**Review by nalimilan — APPROVED** [review]
Thanks!

**Pull request merged**